usvm-ts/src/main/kotlin/org/usvm/TSBinaryOperator.kt

```diff
@@ -6,32 +6,93 @@ sealed class TSBinaryOperator(
+    //
+    // So, that's the reason why banSorts in Neq throws out all primitive types except one of the expressions' one.
     object Neq : TSBinaryOperator(
         onBool = { lhs, rhs -> lhs.neq(rhs) },
         onBv = { lhs, rhs -> lhs.neq(rhs) },
         onFp = { lhs, rhs -> mkFpEqualExpr(lhs, rhs).not() },
+        onRef = { lhs, rhs -> lhs.neq(rhs) },
+        desiredSort = { lhs, _ -> lhs },
+        banSorts = { lhs, rhs ->
+            when {
+                lhs is TSWrappedValue ->
+                    // rhs.sort == addressSort is a mock not to cause undefined
+                    // behaviour with support of new language features.
+                    if (rhs is TSWrappedValue || rhs.sort == addressSort) emptySet() else
+                        TSTypeSystem.primitiveTypes
+                        .map(::typeToSort).toSet()
+                        .minus(rhs.sort)
+                rhs is TSWrappedValue ->
+                    // lhs.sort == addressSort explained as above.
+                    if (lhs.sort == addressSort) emptySet() else
+                        TSTypeSystem.primitiveTypes
+                        .map(::typeToSort).toSet()
+                        .minus(lhs.sort)
+                else -> emptySet()
+            }
+        }
+    )
+
+    object Add : TSBinaryOperator(
+        onFp = { lhs, rhs -> mkFpAddExpr(fpRoundingModeSortDefaultValue(), lhs, rhs) },
+        onBv = UContext<TSSizeSort>::mkBvAddExpr,
+        desiredSort = { _, _ -> fp64Sort },
     )
 
-    internal operator fun invoke(lhs: UExpr<out USort>, rhs: UExpr<out USort>): UExpr<out USort> {
+    object And : TSBinaryOperator(
+        onBool = UContext<TSSizeSort>::mkAnd,
+        desiredSort = { _, _ -> boolSort },
+    )
+
+    internal operator fun invoke(lhs: UExpr<out USort>, rhs: UExpr<out USort>, scope: TSStepScope): UExpr<out USort> {
+        val bannedSorts = lhs.tctx.banSorts(lhs, rhs)
+
+        fun apply(lhs: UExpr<out USort>, rhs: UExpr<out USort>): UExpr<out USort>? {
+            val ctx = lhs.tctx
+            val lhsSort = lhs.sort
+            val rhsSort = rhs.sort
+            assert(lhsSort == rhsSort)
+
+            if (lhsSort in bannedSorts) return null
+            if (ctx.desiredSort(lhsSort, rhsSort) != lhsSort) return null
+
+            return when (lhs.sort) {
+                is UBoolSort -> ctx.onBool(lhs.cast(), rhs.cast())
+                is UBvSort -> ctx.onBv(lhs.cast(), rhs.cast())
+                is UFpSort -> ctx.onFp(lhs.cast(), rhs.cast())
+                is UAddressSort -> ctx.onRef(lhs.cast(), rhs.cast())
+                else -> error("Unexpected sorts: $lhsSort, $rhsSort")
+            }
+        }
+
         val lhsSort = lhs.sort
         val rhsSort = rhs.sort
 
-        if (lhsSort != rhsSort) TODO("Implement type coercion")
+        val ctx = lhs.tctx
+        val sort = ctx.desiredSort(lhsSort, rhsSort)
 
         return when {
```

Comment:
There are no other cases?

Reply:
No, since chosen sort is only used to intersect both exprCaches and have at least one sort to apply binary operation to